### PR TITLE
added PhraseValidator class for checking phrase matching

### PR DESCRIPTION
**Pull request by g-hi3**
I added some code to the `CheckForErrorsMoonscraper` method in `SongValidate` to show errors if phrases are inconsistent (phrase_start and phrase_end do not match). Clone Hero ignores phrases if they're inconsistent. I put the new validation code in the Moonscaper validation because I didn't know how GH3 handles inconsistent phrases.

### CODE REVIEW

**Comment by FireFox2000000** [comment]
If this is for Clone Hero, wouldn't it make more sense to have this in the Clone Hero validation section rather than the Moonscraper one?

**Comment by g-hi3** [comment]
I put the code in the Moonscraper validation, because I assume that GH3 will also behave weirdly with inconsistent secrions. However, I only use Clone Hero, so I can't tell. If you feel it's more appropriate to only validate for Clone Hero, I'll move it there.

**Comment by FireFox2000000** [comment]
Phrases don't exist in GH3 so exclusively Clone Hero would be the better option.

File: Moonscraper Chart Editor/Assets/Scripts/Game/Misc/PhraseValidator.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                    case PhraseEvent.PhraseEnd: {
+                        if (!PhraseEvent.PhraseStart.Equals(previousEvent?.title))
+                        {
+                            validationErrors.Add(new MismatchedPhraseEventError(songEvent));
+                        }
+
+                        break;
+                    }
+                    default:
+                        continue;
```
Comment (FireFox2000000) [inline Moonscraper Chart Editor/Assets/Scripts/Game/Misc/PhraseValidator.cs:36]: This whole loop is a bit hard to read with all the jumping around it does, so could probably do with a slight refactor or some comments but overall this does work so I'm not too fussed about it.

**Review by FireFox2000000 — CHANGES_REQUESTED** [review]
Needs merge conflicts fixed, otherwise approved.

File: Moonscraper Chart Editor/Assets/Scripts/Game/Misc/PhraseValidator.cs (hunk 1 above)
Comment (g-hi3) [reply Moonscraper Chart Editor/Assets/Scripts/Game/Misc/PhraseValidator.cs]: I tried to reduce complexity of the method by reducing the nesting within and being more explicit about what I'm checking.

File: Moonscraper Chart Editor/Assets/Scripts/Game/Misc/SongValidate.cs [hunk 2]
```diff
@@ -1,242 +1,262 @@
...
-        {
-            sb.AppendLine("\tNo errors detected");
-        }
-
-        hasErrors |= hasErrorsLocal;
-
-        return sb.ToString();
-    }
-}
+// Copyright (c) 2016-2020 Alexander Ong
```
Comment (FireFox2000000) [inline Moonscraper Chart Editor/Assets/Scripts/Game/Misc/SongValidate.cs:1]: Something has gone wrong with the line endings on this file. Make sure your git has autocrlf set to true, then try running `git add --renormalize .` and recommitting.

**Review by FireFox2000000 — CHANGES_REQUESTED** [review]

**Review by FireFox2000000 — APPROVED** [review]

**Pull request merged**